autogen/agentchat/contrib/nexusravenv2_local_function_calling.py

```diff
@@ -0,0 +1,162 @@
+
+
+def add_nexus_raven_prompts(func):
+    """
+      ->    messages:[{'content': ' functioncaller.random_word_generator().then(randomWord => mistral.speak(`Using the randomly generated word "${randomWord}," I will now solve this logic problem.`));',
+              'name': 'mistral', 'role': 'user'}],
+              'tools': [
+                      {'type': 'function',
+                      'function': {'description': 'terminate the group chat',
+                                   'name': 'terminate_group_chat',
+                                   'parameters': {'type': 'object',
+                                                  'properties': {'message': {'type': 'string',
+                                                                             'description': 'Message to be sent to the group chat.'}},
+                                                  'required': ['message']}}},
+                     {'type': 'function',
+                     'function': {'description': 'This is a random word generator.',
+                                  'name': 'random_word_generator',
+                                  'parameters': {'type': 'object',
+                                                 'properties': {'seed':   {'type': 'integer',
+                                                                           'description': 'seed for initialising the random generator'},
+                                                                'prefix': {'type': 'string',
+                                                                           'description': 'string to prefix random word'}},
+                                                 'required': ['seed', 'prefix']}}}],
+             'model': 'nexusraven'}]"""
+    if getattr(func, '_is_decorated', False):
+        # If the function is already decorated, return it as is
+        return func
+
+    def wrapper(*args, **kwargs):
+        result = func(*args, **kwargs)
+
+        # extract the tools element from the result
+        query_header = "\n\n".join(
+            [create_nexus_prompt_for_tool(tool) for tool in result['tools'] if tool['type'] == 'function'])
+
+        query = "\n".join([m['content'] for m in result['messages'][-1:]])
+        prompt = f"""<human>:\n{query_header}\nUser Query: Question: {query} \n suggest a function and suitable argument values for the situation conversation:  """
+        del result['messages']
+        del result['tools']
+        result['prompt'] = prompt
+
+        result['extra_query'] = {"inputs": prompt,
+                                 "parameters": {"temperature": 0.001, "do_sample": False, "max_new_tokens": 2000}}
+        result['extra_body'] = {"prompt": prompt}
+        return result
+
+    wrapper._is_decorated = True
+    return wrapper
+
+
+class NexusFunctionCallingAssistant(autogen.AssistantAgent):
+    def __init__(self, **kwargs):
```

Comment:
Since this is a custom agent class for a specific type of model and purpose (i.e., no code execution, etc.), I think we can have a very specialized constructor to expose only model and tool-call specific settings. You might want to customize the system message as well. 

Also, if using custom model client (https://microsoft.github.io/autogen/blog/2024/01/26/Custom-Models), you also want to register the client in the constructor.